### PR TITLE
Introduce SourceChange in place of TextChange

### PR DESCRIPTION
**Pull request by rynowak**
This is the first step on the journey to replacing RazorEditorParser. We
can't just get rid of this because VS is using it today.

### CODE REVIEW

**Comment by rynowak** [comment]
TODO here:
- [ ] I still need to add some unit tests for the new types
- [ ] I need to figure out a few test failures

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs [hunk 1]
```diff
@@ -105,57 +105,57 @@ protected override PartialParseResult CanAcceptChange(Span target, TextChange no
...
         {
             return IsNewDotlessCommitInsertion(target, change) || IsSecondaryDotlessCommitInsertion(target, change);
         }
 
         // Completing 'DateTime' in intellisense with a '.' could result in: '@DateT' -> '@DateT.' -> '@DateTime.' which is accepted.
-        private static bool IsNewDotlessCommitInsertion(Span target, TextChange change)
+        private static bool IsNewDotlessCommitInsertion(Span target, SourceChange change)
         {
             return !IsAtEndOfSpan(target, change) &&
-                   change.NewPosition > 0 &&
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs]: @NTaylorMullen - what's the **meaning** of `NewPosition`?? I just can't get my head around why this code was correct.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs]: There are only two uses of `NewPosition` anywhere, both in this file, and I think that's leading to my 4 test failures.

File: src/Microsoft.AspNetCore.Razor.Language/SourceChange.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+            var offset = GetOffset(span);
+            if (offset == span.Content.Length)
+            {
+                return string.Empty;
+            }
+
+            return span.Content.Substring(offset, span.Length);
+        }
+    }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/SourceChange.cs]: I need to add a `ToString` and equality here also.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/BackgroundParser.cs [hunk 3]
```diff
@@ -67,37 +69,43 @@ protected virtual void OnResultsReady(DocumentParseCompleteEventArgs args)
             }
         }
 
-        internal static bool TreesAreDifferent(RazorSyntaxTree leftTree, RazorSyntaxTree rightTree, IEnumerable<TextChange> changes)
-        {
-            return TreesAreDifferent(leftTree, rightTree, changes, CancellationToken.None);
-        }
-
+#pragma warning disable CS0612 // Type or member is obsolete
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/BackgroundParser.cs]: 😢 we could just align an insertion with WTE on the removal of this

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/ImplicitExpressionEditHandler.cs]: Absolutely no clue

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/AutoCompleteEditHandler.cs [hunk 4]
```diff
@@ -31,11 +31,12 @@ public AutoCompleteEditHandler(Func<string, IEnumerable<ISymbol>> tokenizer, Acc
...
         public string AutoCompleteString { get; set; }
 
-        protected override PartialParseResult CanAcceptChange(Span target, TextChange normalizedChange)
+        protected override PartialParseResult CanAcceptChange(Span target, SourceChange change)
         {
-            if (((AutoCompleteAtEndOfSpan && IsAtEndOfSpan(target, normalizedChange)) || IsAtEndOfFirstLine(target, normalizedChange)) &&
-                normalizedChange.IsInsert &&
-                ParserHelpers.IsNewLine(normalizedChange.NewText) &&
+            if (((AutoCompleteAtEndOfSpan && IsAtEndOfSpan(target, change)) || IsAtEndOfFirstLine(target, change)) &&
+                change.Span.Length == 0 &&
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/AutoCompleteEditHandler.cs]: Why not just use IsInsert on the source change?

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs [hunk 5]
```diff
@@ -82,42 +82,39 @@ protected virtual SpanBuilder UpdateSpan(Span target, TextChange normalizedChang
             return newSpan;
         }
 
-        protected internal static bool IsAtEndOfFirstLine(Span target, TextChange change)
+        public string ApplyChange(SourceChange change, string content, int changeOffset)
+        {
+            var changeRelativePosition = change.Span.AbsoluteIndex - changeOffset;
+
+            Debug.Assert(changeRelativePosition >= 0);
+            return content.Remove(changeRelativePosition, change.Span.Length).Insert(changeRelativePosition, change.NewText);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs]: Could we make this bit more debuggable. Have a feeling we might be debugging through here in the future:
```C#
var appliedChange = content.Remove(changeRelativePosition, change.Span.Length).Insert(changeRelativePosition, change.NewText);
return appliedChange;
```

File: src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Razor
+{
+    internal static class TextChangeExtensions
+    {
+        public static SourceChange AsSourceChange(this TextChange textChange)
```
Comment (NTaylorMullen) [inline src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs]: Why doesn't this guy need a pragma warn?

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/BackgroundParser.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/BackgroundParser.cs]: I don't want to make incremental updates to `RazorEditorParser`, I'm trying to replace it completely.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/AutoCompleteEditHandler.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/AutoCompleteEditHandler.cs]: Yeah I will. I think I wrote this before I added `IsInsert`

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs]: Sure

File: src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs]: hmm wierd.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/SpanEditHandler.cs]: Actually going to remove this, this should be part of `SourceChange`

File: src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor/TextChangeExtensions.cs]: Ok. Because this is the Roslyn type, not the Razor legacy type.

**Comment by rynowak** [comment]
Updated

**Pull request merged**